source/common/config/unified_mux/subscription_state.h

```diff
@@ -0,0 +1,106 @@
+        ttl_([this](const std::vector<std::string>& expired) { ttlExpiryCallback(expired); },
+             dispatcher, dispatcher.timeSource()),
+        type_url_(std::move(type_url)), callbacks_(callbacks), dispatcher_(dispatcher) {
+    if (init_fetch_timeout.count() > 0 && !init_fetch_timeout_timer_) {
+      init_fetch_timeout_timer_ = dispatcher.createTimer([this]() -> void {
+        ENVOY_LOG(warn, "config: initial fetch timed out for {}", type_url_);
+        callbacks_.onConfigUpdateFailed(ConfigUpdateFailureReason::FetchTimedout, nullptr);
+      });
+      init_fetch_timeout_timer_->enableTimer(init_fetch_timeout);
+    }
+  }
+
+  virtual ~SubscriptionState() = default;
+
+  // Update which resources we're interested in subscribing to.
+  virtual void updateSubscriptionInterest(const absl::flat_hash_set<std::string>& cur_added,
+                                          const absl::flat_hash_set<std::string>& cur_removed) PURE;
+
+  void setDynamicContextChanged() { dynamic_context_changed_ = true; }
+  void clearDynamicContextChanged() { dynamic_context_changed_ = false; }
+  bool dynamicContextChanged() const { return dynamic_context_changed_; }
+
+  // Whether there was a change in our subscription interest we have yet to inform the server of.
+  virtual bool subscriptionUpdatePending() const PURE;
+
+  virtual void markStreamFresh() PURE;
+
+  // Implementations expect either a DeltaDiscoveryResponse or DiscoveryResponse. The caller is
+  // expected to know which it should be providing.
+  virtual UpdateAck handleResponse(const RS& response_proto_ptr) PURE;
+
+  virtual void handleEstablishmentFailure() PURE;
+
+  // Returns the next gRPC request proto to be sent off to the server, based on this object's
+  // understanding of the current protocol state, and new resources that Envoy wants to request.
+  // Returns a new'd pointer, meant to be owned by the caller, who is expected to know what type the
+  // pointer actually is.
+  virtual RQ* getNextRequestAckless() PURE;
+  // The WithAck version first calls the ack-less version, then adds in the passed-in ack.
+  // Returns a new'd pointer, meant to be owned by the caller, who is expected to know what type the
+  // pointer actually is.
+  virtual RQ* getNextRequestWithAck(const UpdateAck& ack) PURE;
+
+  void disableInitFetchTimeoutTimer() {
+    if (init_fetch_timeout_timer_) {
+      init_fetch_timeout_timer_->disableTimer();
+      init_fetch_timeout_timer_.reset();
+    }
+  }
+
+  virtual void ttlExpiryCallback(const std::vector<std::string>& type_url) PURE;
+
+protected:
+  std::string type_url() const { return type_url_; }
+  UntypedConfigUpdateCallbacks& callbacks() const { return callbacks_; }
+
+  // Not all xDS resources supports heartbeats due to there being specific information encoded in
+  // an empty response, which is indistinguishable from a heartbeat in some cases. For now we just
```

Comment:
Not sure if I'm missing something here. An empty response in SotW mode is very different than an empty response in Delta mode, no?
In that case, could it be that this field (`supports_heartbeats_`) is only relevant for Delta?

Reply:
Nods, it's Delta specific and it's doubtful this will change; there was no reason to put the flag in the base class really. Moved to delta state.